File: Algorithms/Python3.x/111-Minimum_Depth_of_Binary_Tree.py
# Runtime: 56 ms, faster than 29.51% of Python3 online submissions for Minimum Depth of Binary Tree.
# Memory Usage: 15.8 MB, less than 29.87% of Python3 online submissions for Minimum Depth of Binary Tree.
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if root is None:
            return 0
        if root.left is None and root.right is None:
            return 1
        left = self.minDepth(root.left) + 1 if root.left is not None else float("inf")
        right = self.minDepth(root.right) + 1 if root.right is not None else float("inf")
        return min(left, right)


# Runtime: 48 ms, faster than 82.57% of Python3 online submissions for Minimum Depth of Binary Tree.
# Memory Usage: 14.4 MB, less than 91.69% of Python3 online submissions for Minimum Depth of Binary Tree.
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def minDepth(self, root):
        if not root:
            return 0
        queue = [(root, 1)]
        while queue:
            node, level = queue.pop(0)
            if node:
                if node.left is None and node.right is None:
                    return level
                else:
                    queue.append((node.left, level + 1))
                    queue.append((node.right, level + 1))

File: Algorithms/C++/111-Minimum_Depth_of_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        // 1.左孩子和有孩子都为空的情况，说明到达了叶子节点，直接返回1即可
        if (root == nullptr) return 0;
        if (root->left == nullptr && root->right == nullptr) return 1;
        // 2.如果左孩子和由孩子其中一个为空，那么需要返回比较大的那个孩子的深度      
        int minLeft = minDepth(root->left),
            minRight = minDepth(root->right);
        // 这里其中一个节点为空，说明minLeft和minRight有一个必然为0，所以可以返回m1 + m2 + 1;
        if (root->left == nullptr || root->right == nullptr) return minLeft + minRight + 1;
        // 3.最后一种情况，也就是左右孩子都不为空，返回最小深度+1即可
        return std::min(minLeft, minRight) + 1;
    }
};
